File: P101SymmetricTree/symmetric_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    """
    Returns whether a binary tree is symmetric or not.

    A binary tree is symmetric if there exists a plane of symmetry when
    a line is drawn vertically from the root through the center to the 
    last level of the tree.

    Parameters
    ----------
    root : TreeNode
        Root node of the binary tree.

    Returns
    -------
    is_symmetric : bool
        Whether the binary tree is symmetric or not.

    Examples
    --------
    Input : [1,2,2,3,4,4,3]
              1
             / \
            2   2
           / \ / \
          3  4 4  3
    Output : True
 
    Input : [1,2,2,null,3,null,3]
              1
             / \
            2   2
             \   \
             3    3
    Output : True
    
    Input : [1,17,17,5,null,null,5]
              1
             / \
           17   17
           /     \
          5       5
    Output : True
    """
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True

        def is_symmetric_helper(node_a, node_b) -> bool:
            if node_a is None and node_b is None:
                return True
            elif node_a is not None and node_b is None: 
                return False
            elif node_b is not None and node_a is None:
                return False

            if node_a.val == node_b.val:
                return is_symmetric_helper(node_a.left,  node_b.right) and \
                       is_symmetric_helper(node_a.right, node_b.left)

            return False

        return is_symmetric_helper(root.left, root.right)

File: P101SymmetricTree/symmetric_tree.cxx
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    /**
     * @brief Recursive helper function which determines if a binary \
     *  tree is symmetric or not.
     * 
     * @param a TreeNode*
     * @param b TreeNode*
     * 
     */
    bool is_symmetric_helper(TreeNode *a, TreeNode *b)
    {
        // base case, pointers have recursed all the way down and
        // both left and right tree nodes are NULL
        if (a == nullptr && b == nullptr) {
            return true;
        }

        if ((a == nullptr && b != nullptr) || 
            (a != nullptr && b == nullptr)) {
            return false;
        }

        if (a->val == b->val) {
            return is_symmetric_helper(a->left, b->right) && 
                   is_symmetric_helper(a->right, b->left);
        } 
        
        return false;
    }

    /**
     * @brief Returns whether a binary tree is symmetric or not.
     *
     * A binary tree is symmetric if there exists a plane of symmetry when
     * a line is drawn vertically from the root through the center to the 
     * last level of the tree.
     * 
     * @param a TreeNode*
     * @param b TreeNode*
     */
    bool isSymmetric(TreeNode* root)
    {
        if (root == nullptr)
            return true;

        return is_symmetric_helper(root->left, root->right);    
    }
};